### PR TITLE
Drive train encoders

### PR DESCRIPTION
**Pull request by jameshochadel**
Submitting PR to do code review as requested in https://trello.com/c/gVrBUjBK

### CODE REVIEW

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Do we need to set each of these if leftMiddleTalon and leftBackTalon are in Follow mode?

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Do we need to set each of these if leftMiddleTalon and leftBackTalon are in Follow mode?

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Do we need to set each of these if leftMiddleTalon and leftBackTalon are in Follow mode?

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Do we need to set each of these if leftMiddleTalon and leftBackTalon are in Follow mode?

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: The drive*Velocity methods and drive*Percent methods both expect a double, but we know that they will be doubles in different ranges (e.g. Percent should get a double in range 0-1, while Velocity should get one in a larger range). Should these methods test for that somehow?

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: DRIVETRAIN_WHEEL_DIAMETER_CONSTANT is set to 0 in RobotMap, so this constant also evaluates to 0; thus, this method (and getRightPosition) will always return 0

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: This is nitpicky, but there's no need for the extra whitespace

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Is this the right name for a method that returns the P-constant? (Genuine question, as I am not sure)

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: These need to be set before this can be merged

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: This should be resolved before merge

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: TODO should be resolved before merge

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: Actual value needs to be set

File: src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java]: Not your fault, but this is never used

File: src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java]: Naming: Should this be setDriveAngle?

File: src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java]: Redundancy: Does this need set if it's already set in the constructor?

**Review by jameshochadel — COMMENTED** [review]
Review complete; needs work

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (wmarshall) [reply src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Or at least document the units in JavaDoc

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (wmarshall) [reply src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: We should ask Design what the wheel diameter is.

We should also document what unit(inches?) this returns

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: This does not need to be completed for this review to pass, BUT I made a work item for experimentally determining these: https://trello.com/c/06T2jY5o

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (wmarshall) [reply src/org/usfirst/frc/team2399/robot/RobotMap.java]: Still needs to be set

File: src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java
Comment (wmarshall) [inline src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java]: What is moveToAngle() doing here?

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (wmarshall) [reply src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Still does not document units. Soft limits are good though. Use the @param tag to talk about it

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (wmarshall) [reply src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Persnickety thing: We know it's a constant because it came from RobotMap and is all caps. the _CONSTANT is redundant.

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: Sanity check: For robot to go backwards, lower limit must be -1, yes?

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: You mention this in the todo, but we need to talk to Design to get these before this is ready for merge

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (wmarshall) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Why are we keeping an internal currentPConstant and querying the talon with getP() every call?

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (wmarshall) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Please resolve before merge

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Two issues with this one:
1. The method is currently a stub, so I don't see why it should exist at all if it's empty.
2. If this is intended for what I believe it is— rotating the robot in the xy plane— it should not be in the Subsystem. A `Subsystem` is a programmatic representation of the physical robot subsystem, and is meant only to provide methods for directly controlling that subsystem: For example, "make these motors spin at this RPM". In other words, it establishes a programmatic view into **what is possible with the subsystem**. This method is better suited for a `Command`: It represents a task that we want the robot to do, and instructs the `DriveTrain` on how to do it.

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: New rule, which I should've mentioned before: Constants should not end in _CONSTANT. The fact that they're capitalized and in RobotMap tells us what they are; the extra _CONSTANT suffix is redundant. Please rename all appropriate constants.

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java:55]: Are -1 and 1 the actual speeds, or does this still need resolved?

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: Just noticed, "decrement" is misspelled

File: src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java]: Can you clarify what this means?

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: My eclipse marks these two imports as unused

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Eclipse marks this and the below field as unused

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (Num3ra1s) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: Done

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java:68]: Eclipse finds no references to this constant in the project. Why do we have it?

File: src/org/usfirst/frc/team2399/robot/RobotMap.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: I believe that the "@param" annotation is meant for parameters to methods. For constants like these, you can just list it in a generic javadoc, e.g. 
`/**
  your description here
  */`

File: src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/commands/JoyDrive.java]: As usual, minor thing, but everyone's Eclipse should be set to automatically remove trailing whitespace on save. `git blame` tells me this one was Andrew

File: src/org/usfirst/frc/team2399/robot/subsystems/Climber.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/Climber.java]: Dat capitalized CLimber though

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Minor naming quibble: Are we "calculating" anything, or are we just getting it? (I'm open to be convinced either way)

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Hnnng

**Review by jameshochadel — COMMENTED** [review]
Final review! A few very minor style issues and we'll be ready to merge. Good job, everyone.

**Review by chloe6248 — APPROVED** [review]
Got approval from James :)

**Pull request merged**
